### PR TITLE
Reports Milestone 1 implementation

### PR DESCRIPTION
**Pull request by sumanvpacewisdom**

### CODE REVIEW

File: src/constants/common.js [hunk 1]
```diff
@@ -119,6 +119,8 @@ module.exports = {
 	getDefaultOrgPolicies: defaultOrgPolicies,
 	CURRENT: 'CURRENT',
 	ALL: 'ALL',
+	All: 'All',
```
Comment (rakeshSgr) [inline src/constants/common.js:122]: @sumanvpacewisdom we can use the above one Why one more needed.  ?

File: src/controllers/v1/reports.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	/**
+	 * Report filter list
+	 * @method
+	 * @name reportFilterList
+	 * @param {Object} req - request data.
+	 * @param {String} req.decodedToken.token - user token.
+	 * @returns {JSON} - filter list.
+	 */
+
+	async reportFilterList(req) {
```
Comment (rakeshSgr) [inline src/controllers/v1/reports.js:13]: @sumanvpacewisdom controller itself is report, Just name the function as filterList

Comment (sumanvpacewisdom) [reply src/controllers/v1/reports.js:13]: @rakeshSgr its for label and value

File: src/controllers/v1/reports.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+const reportService = require('@services/reports')
+
+module.exports = class Reports {
+	/**
+	 * Report filter list
+	 * @method
+	 * @name reportFilterList
+	 * @param {Object} req - request data.
+	 * @param {String} req.decodedToken.token - user token.
+	 * @returns {JSON} - filter list.
```
[the remark this answers is not in the document]
Comment (sumanvpacewisdom) [reply src/controllers/v1/reports.js:10]: ok

File: src/database/migrations/20241205081137-add_report_role_maping_table.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('report_role_mapping', {
+			id: {
+				allowNull: false,
+				autoIncrement: true,
+				primaryKey: true,
+				type: Sequelize.INTEGER,
+			},
+			report_id: {
```
Comment (rakeshSgr) [inline src/database/migrations/20241205081137-add_report_role_maping_table.js:13]: report code we can store here

File: src/database/migrations/20241205082532-add_reports_table.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('reports', {
+			id: {
+				allowNull: false,
+				autoIncrement: true,
+				primaryKey: true,
+				type: Sequelize.INTEGER,
+			},
+			code: {
+				allowNull: false,
```
Comment (rakeshSgr) [inline src/database/migrations/20241205082532-add_reports_table.js:14]: @sumanvpacewisdom set code as unique

File: src/database/migrations/20241205082735-add_report_types_table.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('report_types', {
+			id: {
+				allowNull: false,
+				autoIncrement: true,
+				primaryKey: true,
+				type: Sequelize.INTEGER,
+			},
+			title: {
+				type: Sequelize.STRING,
```
Comment (rakeshSgr) [inline src/database/migrations/20241205082735-add_report_types_table.js:14]: @sumanvpacewisdom set this as unique

File: src/database/migrations/20241212202720-add_role_extensions_data.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+					scope: 'PUBLIC',
+					organization_id: '1',
+					created_at: new Date(),
+					updated_at: new Date(),
+					deleted_at: null,
+				},
+				{
+					title: 'mentee',
+					label: 'Mentee',
+					status: 'INACTIVE',
```
Comment (rakeshSgr) [inline src/database/migrations/20241212202720-add_role_extensions_data.js:34]: @sumanvpacewisdom why this is inactive

File: src/database/migrations/20241212202720-add_role_extensions_data.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+require('module-alias/register')
+const common = require('@constants/common')
+
+module.exports = {
+	up: async (queryInterface, Sequelize) => {
+		try {
+			// Data to be inserted
+			const roleExtensionsData = [
+				{
+					title: 'user',
```
Comment (rakeshSgr) [inline src/database/migrations/20241212202720-add_role_extensions_data.js:12]: @sumanvpacewisdom do we have this role in user service ?

File: src/database/migrations/20241205073509-add_role_extentions_table.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+'use strict'
+/** @type {import('sequelize-cli').Migration} */
+
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('role_extensions', {
```
Comment (rakeshSgr) [inline src/database/migrations/20241205073509-add_role_extentions_table.js:6]: @sumanvpacewisdom add id column  and set as auto numeric

File: src/controllers/v1/reports.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sumanvpacewisdom) [reply src/controllers/v1/reports.js:10]: @rakeshSgr i don't find it anywhere giving example of the json filter list, please let me know example how it should be done

File: src/database/migrations/20241212202720-add_role_extensions_data.js (hunk 8 above)
Comment (sumanvpacewisdom) [reply src/database/migrations/20241212202720-add_role_extensions_data.js:12]: @rakeshSgr 
yes we have in users services , we are using it in the bulkuserupload

File: src/services/reports.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	 * Get entityTypes for reports
+	 * @method
+	 * @name getReportFilterList
+	 * @param {String} tokenInformation - token information
+	 * @param {Boolean} queryParams - queryParams
+	 * @returns {JSON} - Report filter list.
+	 */
+	static async getReportFilterList(organization, entity_type, filterType, tokenInformation) {
+		try {
+			const reportFilterList = await menteesService.getFilterList(
```
Comment (rakeshSgr) [inline src/services/reports.js:17]: @sumanvpacewisdom we need to get only entity which have report flag set as true

File: src/constants/common.js (hunk 1 above)
Comment (rakeshSgr) [reply src/constants/common.js:122]: make is as all_label

File: src/database/migrations/20241205073509-add_role_extentions_table.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+			id: {
+				allowNull: false,
+				autoIncrement: true,
+				primaryKey: true,
```
Comment (rakeshSgr) [inline src/database/migrations/20241205073509-add_role_extentions_table.js:10]: @sumanvpacewisdom remove primary key for this

File: src/database/migrations/20241205082532-add_reports_table.js (hunk 5 above)
Comment (rakeshSgr) [reply src/database/migrations/20241205082532-add_reports_table.js:14]: @sumanvpacewisdom set this is a primary and refer this in other tables like report queries as well

File: src/database/migrations/20241205082907-add_report_queries_table.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('report_queries', {
+			id: {
+				allowNull: false,
+				autoIncrement: true,
+				primaryKey: true,
+				type: Sequelize.INTEGER,
+			},
+			report_id: {
```
Comment (rakeshSgr) [inline src/database/migrations/20241205082907-add_report_queries_table.js:13]: @sumanvpacewisdom use report_code

File: src/database/migrations/20241212203721-add-role-permissions-for-reports.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+	}
+}
+
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		try {
+			const rolePermissionsData = [
+				{
+					role_title: common.ORG_ADMIN_ROLE,
+					permission_id: await getPermissionId('reports', ['GET'], '/mentoring/v1/reports/filterList'),
```
Comment (rakeshSgr) [inline src/database/migrations/20241212203721-add-role-permissions-for-reports.js:29]: @sumanvpacewisdom these api's can be accessed by the mentee,mentor, org admin and session manager. 
Provide permission for those roles as well.

File: src/database/migrations/20241216110839-update-enitityType-report-filter.js [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+				{ value: 'type', label: 'Type' },
+				{ value: 'recommended_for', label: 'Recommended For' },
+				{ value: 'categories', label: 'Categories' },
+			]
+
+			// Fetch and update records for each value/label pair
+			const updateEntityTypePromises = entityTypesToUpdate.map(async (item) => {
+				try {
+					const entityTypes = await queryInterface.sequelize.query(
+						`SELECT id, value, label, organization_id, model_names
```
Comment (rakeshSgr) [inline src/database/migrations/20241216110839-update-enitityType-report-filter.js:21]: @sumanvpacewisdom is it need to select these columns for verifying the entity type is exist or not ?

File: src/database/models/reportQueries.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+	const ReportQuery = sequelize.define(
+		'ReportQuery',
+		{
+			id: {
+				type: DataTypes.INTEGER,
+				allowNull: false,
+				primaryKey: true,
+				autoIncrement: true,
+			},
+			report_id: {
```
Comment (rakeshSgr) [inline src/database/models/reportQueries.js:13]: @sumanvpacewisdom use report_code

File: src/database/models/reportRoleMapping.js [hunk 16]
```diff
@@ -0,0 +1,45 @@
+'use strict'
+
+module.exports = (sequelize, DataTypes) => {
+	const ReportRoleMapping = sequelize.define(
+		'ReportRoleMapping',
+		{
```
Comment (rakeshSgr) [inline src/database/models/reportRoleMapping.js:6]: @sumanvpacewisdom make the same schema changes which mentioned for the migration.

File: src/services/reports.js [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+		} catch (error) {
+			return {
+				success: false,
+				message: error.message,
+			}
+		}
+	}
+}
+
+function convertEntitiesForFilter(entityTypes) {
```
Comment (rakeshSgr) [inline src/services/reports.js:315]: @sumanvpacewisdom as the same code used in meentes.js as well, move this utils and make changes in meentes.js as well

File: src/services/reports.js [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+			organization_id: entityType.organization_id,
+			entities: entityType.entities || [],
+		}
+
+		result[key].push(newObj)
+	})
+	return result
+}
+
+function filterEntitiesBasedOnParent(data, defaultOrgId, doNotRemoveDefaultOrg) {
```
Comment (rakeshSgr) [inline src/services/reports.js:339]: @sumanvpacewisdom don't duplicate the code in multiple places, move it one file use it in both relevant functions

File: src/helpers/getOrgIdAndEntityTypewithEntitiesBasedOnPolicy.js [hunk 19]
```diff
@@ -201,7 +200,9 @@ module.exports = class OrganizationAndEntityTypePolicyHelper {
 					[Op.in]: entityTypes,
 				}
 			}
-
+			if (report_filter) {
```
Comment (rakeshSgr) [inline src/helpers/getOrgIdAndEntityTypewithEntitiesBasedOnPolicy.js:203]: @sumanvpacewisdom pass the filter from the calling function and append directly to the filter, remove this condition

File: src/services/reports.js [hunk 20]
```diff
@@ -24,7 +24,7 @@ module.exports = class ReportsHelper {
 			}
 
 			const filter_type = filterType !== '' ? filterType : common.MENTOR_ROLE
-			const report_filter = reportFilter !== '' ? reportFilter : true
+			const report_filter = reportFilter
```
Comment (rakeshSgr) [inline src/services/reports.js:27]: @sumanvpacewisdom pass the filter.report_filter = true/ false from here

**Pull request merged**